### ISSUE TITLE
Introduce new type of custom project setting: file path

### ISSUE THREAD

**Issue by miskatonicstudio** [godotengine/godot-proposals#3620]
### Describe the project you are working on

I'm working on GOAT (Godot Open Adventure Template):
https://github.com/miskatonicstudio/goat

### Describe the problem or limitation you are having in your project

The settings of my framework are currently configured in global files. It would be more elegant to store them as custom Project Settings. That seems to work very well, but some of GOAT settings are actually paths to files (similar to e.g. "Main Scene" in Application=>Run settings). However, there is no "file path" type to choose from, so I have to use String. That works, but providing a file path hint (along with a popup to choose a file from all project files, similar to "Main Scene") would work much better ;)

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Provide a new type of custom Project Settings type: `File` (or `File Path`, or `Path`). It would be stored as a String, but that string would always start with `res://` and would represent a path to a file (within a project). It would also provide the same type of file selection popup as "Main Scene" setting.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

AFAIK "Main Scene" is also stored as a String, but it starts with `res://`. The difference is that is uses `PROPERTY_HINT_FILE`. This new type should work in exactly this way. In case of an empty value, the `project.godot` file could store just "res://". In short, I would like this setting to work (from the user's perspective) exactly like "Main Scene" in Application=>Run. It is just a matter of providing a proper hint for the string.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

In GDScript: I don't think so, this is Project Settings level. However, I believe that reusing the code from default Project Settings should not be difficult, as it already works for some default settings (e.g. for "Main Scene"). It would just be a matter of enabling that setting type (more precisely: String type with a proper hint) to custom settings.

### Is there a reason why this should be core and not an add-on in the asset library?

Default Godot settings provide this type of variables already, I think this should be extended to custom settings as part of the core functionality ;)

**Comment by Calinou** [comment]
You can already specify custom property hint info as follows: https://github.com/Calinou/godot/blob/5acb8a253e40c8e8f86c2e58f9cc7e1a9a020ccc/core/config/project_settings.cpp#L1206

The above code is C++, but you can achieve the same task using the `ProjectSettings.add_property_info()` method from a script.

As for supplying an array or dictionary of file paths, this is already being tracked in https://github.com/godotengine/godot-proposals/issues/3502.

**Comment by miskatonicstudio** [comment]
Thanks for a fast reply! Is it possible to set the hint using GDScript in a way that persists in Project Settings? The whole point of moving the settings from a global file (AutoLoad) to Project Settings is so that they are easily available (with proper types) and also available before any script is executed (e.g. `GAME_RESOURCES_DIRECTORY` is used to look for voice, inventory items etc. across multiple global scripts).

**Comment by Calinou** [comment]
> Is it possible to set the hint using GDScript in a way that persists in Project Settings? The whole point of moving the settings from a global file (AutoLoad) to Project Settings is so that they are easily available (with proper types) and also available before any script is executed (e.g. `GAME_RESOURCES_DIRECTORY` is used to look for voice, inventory items etc. across multiple global scripts).

The property hint does not modify the underlying property's type – it's still a string.

**Comment by miskatonicstudio** [comment]
What I meant is that I'd like to have a setting, available in Project Settings => Goat, that shows a nice popup to select a file/directory. It should work exactly like "Main Scene" and be available in the same way, directly in Project Settings. Is such a thing possible now, using the solution you've mentioned and GDScript?

**Comment by Calinou** [comment]
> What I meant is that I'd like to have a setting, available in Project Settings => Goat, that shows a nice popup to select a file/directory. It should work exactly like "Main Scene" and be available in the same way, directly in Project Settings. Is such a thing possible now, using the solution you've mentioned and GDScript?

Yes, you can create project settings with the path `goat/some_setting_name` using `ProjectSettings.set()`, then use `ProjectSettings.add_property_info()` to add custom hints. They will appear in a Goat section in the Project Settings.

**Comment by miskatonicstudio** [comment]
I've added a script to one of the singletons (AutoLoad):

```
func _ready():
	ProjectSettings.set("goat/common/TEST", 1)
	var property_info = {
		"name": "goat/common/TEST",
		"type": TYPE_INT,
		"hint": PROPERTY_HINT_ENUM,
		"hint_string": "one,two,three"
	}
	ProjectSettings.add_property_info(property_info)
	ProjectSettings.save()
```

Then I started the program and exited. I was asked if I want to reload `project.godot`, which I did. The property was set, with a correct value, but without any hint:

![Screenshot from 2021-12-02 15-55-36](https://user-images.githubusercontent.com/36821133/144446185-887f54a5-da8a-44f6-bd78-8ea227f0fc90.png)

The same happened with `PROPERTY_HINT_FILE`. Could you please advise on where the script should be placed/how it should be executed?

The goal here is to open Project Settings in Godot and see that property in exactly the same way as Main Scene, with a small file icon that opens a file selection popup:

![Screenshot from 2021-12-02 15-57-38](https://user-images.githubusercontent.com/36821133/144446590-c60f9fa6-5008-4c3e-bd4d-ffd03f7a606d.png)

**Comment by Calinou** [comment]
The property hint needs to be set in the editor itself, so you need to do that from a `tool` script.

**Comment by miskatonicstudio** [comment]
OK, I've created a `tool` script that is also a singleton and it seems to work exactly as expected! :) Thank you for explaining the proper workflow to me @Calinou :+1: 

For GOAT purpose, I'd say this solves my issue. However, do you think we could still have this as part of the core functionality? That was a lot of extra steps I had to take to display a nice directory picker popup. Maybe this could be configurable when a property is added instead, to keep all Project Settings related properties in one place?

**Comment by Calinou** [comment]
> However, do you think we could still have this as part of the core functionality? That

It can't really be done any other way, unless we add a dedicated FilePath type (similar to NodePath). However, I feel this is overkill for something that will only be used within the editor.
